### PR TITLE
Keyname overhaul

### LINKED ISSUE #2698: arrays with natural sorting
As discussed in #182 and #2388 we could avoid the `_` for array numbers if we internally escape `#<nr>` to `#_..<nr>` (elektraUnescapeKeyName):

> As already discussed shortly, we could also extend the escaping logic for key names and avoid the _ prefixes. So it would be #10, and the (internally) unescaped version (see keyUnescapedName) would be #_10 for correct sorting.

Then the sorting would be still correct and the user would not be bothered by `___` (as mentioned recently in #2685)

Furthermore, we maybe should change the character for arrays to not be `#` as `#` is a shell comment (see #2388).

### LINKED ISSUE #3050: remove elektraKeySetName
The idea of elektraKeySetName was to have flags which change the restrictions on the names (to allow cascading, meta, and empty key names). Now:

1. cascading names are allowed by default
2. against empty key names we are already protected by `ksAppendKey`
3. against meta/non-meta key names we will get protection with #2983

So we actually do not need the argument for elektraKeySetName. So I propose to remove elektraKeySetName and always allow every name. The introspection which key name was used can be done with keyGetNamespace (see also #3049).

This change has effect on tooling: the error message will not be thrown at `keySetName` but later at `ksAppendKey`.

### LINKED ISSUE #3082: default namespace
As already discussed in #2723, #2983 and others (older issues I could not find about the spec plugin) it would be clearer if we would not abuse cascading keys to also represent default keys.

## Current Situation

Currently, the spec plugin creates cascading keys for all `default` it does find. `ksLookup` uses these keys (instead of creating new cascading keys, see #972) if they are found.

## Wanted Situation

Not cascading keys but `default:` keys should be added in the spec plugin (and within `ksLookup` if we cannot fix #972)

## Advantages

Following advantages are expected:

- ksAppendKey could reject cascading keys (only default keys would be allowed)
- The `default` key would make clearly visible what kind of key it is (not clear if `cascading` keys are reused for different purposes.

### LINKED ISSUE #3084: order of namespaces in KeySet
Currently keys are added in alphabetical order, thus the namespaces would be in following order:

1. default
1. dir
1. proc
1. spec
1. system
1. user

More logical, however, would be to have the keys in order of preference:

1. spec
1. proc
1. dir
1. user
1. system
1. default

This could be implemented similar to #2698: Within the unescaped key names there could be some prefix that brings the keys to the desired order.

### PR DESCRIPTION
**Pull request by kodebach**
Continuation of #3447 (commits squashed and rebased)

Implements #3102

Closes #3050
Closes #3084
Closes #2698
Closes #3082

## Basics

These points need to be fulfilled for every PR:

- [ ] Short descriptions of your changes are in the release notes
      (added as entry in `doc/news/_preparation_next_release.md` which
      contains `_(my name)_`)
      **Please always add something to the release notes.**
- [ ] Details of what you changed are in commit messages
      (first line should have `module: short statement` syntax)
- [ ] References to issues, e.g. `close #X`, are in the commit messages.
- [ ] The buildservers are happy.
- [ ] The PR is rebased with current master.

If you have any troubles fulfilling these criteria, please write
about the trouble as comment in the PR. We will help you.
But we cannot accept PRs that do not fulfill the basics.

## Checklist

Check relevant points but **please do not remove entries**.
For docu fixes, spell checking, and similar none of these points below
need to be checked.

- [ ] I added unit tests for my code
- [ ] I fully described what my PR does in the documentation
      (not in the PR description)
- [ ] I fixed all affected documentation
- [ ] I added code comments, logging, and assertions as appropriate (see [Coding Guidelines](https://master.libelektra.org/doc/CODING.md))
- [ ] I updated all meta data (e.g. README.md of plugins and [METADATA.ini](https://master.libelektra.org/doc/METADATA.ini))
- [ ] I mentioned every code not directly written by me in [THIRD-PARTY-LICENSES](doc/THIRD-PARTY-LICENSES)

## Review

Reviewers will usually check the following:

- [ ] Documentation is introductory, concise, good to read and describes everything what the PR does
- [ ] Examples are well chosen and understandable
- [ ] Code is conforming to [our Coding Guidelines](https://master.libelektra.org/doc/CODING.md)
- [ ] APIs are conforming to [our Design Guidelines](https://master.libelektra.org/doc/DESIGN.md)
- [ ] Code is consistent to [our Design Decisions](https://master.libelektra.org/doc/decisions)

### CODE REVIEW

**Comment by kodebach** [comment]
@markus2330 Please check, if the new Python reference implementation and the documentation matches the decisions in #3549.

The changes can be found in this commit: ae05d3191652e3dbbbd7c77cf8dd62c034130bf0 

If everything matches, I will update the C implementation so that this PR can be merged ASAP. Then we can start on the other API changes (`keyNew`, removal of escaped name, etc.) in separate PRs.

**Comment by markus2330** [comment]
I looked at ae05d31, I would simplify it and give away this special handling about `#` (make it pure conventional).

**Comment by markus2330** [comment]
The part with . and .. looks perfect to me now! :1st_place_medal:

**Comment by kodebach** [comment]
> I looked at ae05d31, I would simplify it and give away this special handling about `#` (make it pure conventional).

I already explained it a bit, but here is a more in depth explanation:

1. We decided that we want a 1:1 relation between escaped and unescaped forms.
2. We also want that `#123` works as an array index, i.e. it turns into `#__123` automatically.

Just based 2 we already need at least some form of escaping mechanism, or `keyAddBaseName (k, "#123")` results in a Key that has no escaped name. We could just use `.../#123` as the escaped name and say that you can construct such a name only via `keyAddBaseName`. But that has another problem. It is very common to do `keyNew (keyName(k), KEY_END)` to create a Key with the same name as `k`, but nothing else in common. This wouldn't work anymore, since it would result in a key with name `.../#__123`.

To have the 1:1 relation the `\#` can only be valid, when `#` has an effect. Otherwise we would have some unescaped names with two escaped forms. For example `\#abc` and `#abc` would both be unescaped into `#abc`. This can be solved, making either `\#abc` or `#abc` illegal in the escaped name. Here I decided to make the first one illegal. In the previous version `#abc` was illegal, but we decided that was to much "semantics".

I actually just noticed that the check for `\#` is not correct. `\#` should only be allowed for the case were `_` will be inserted. So line 90 should actually be the much simpler:

```python
"#": lambda part: re.match(r"^\\#[1-9][0-9]{1,18}$", part) and (len(part) < 21 or part[2:] <= "9223372036854775807"),
```

**Comment by markus2330** [comment]
Thank you, I was just missing what you write here in the docu! With this fixed the commit looks fine.

Is there any other specific part which needs a closer look?

**Comment by kodebach** [comment]
> Is there any other specific part which needs a closer look?

No, apart from ae05d31 everything is unchanged compared to #3447. I'll fix the Python code and add the details about arrays and then I'll start updating the C code. With a bit of luck this PR can be merged by the end of the week.

**Comment by kodebach** [comment]
@mpranj This PR should finally be ready this weekend. It would be nice, if don't merge any PRs until the meeting on Monday (too avoid more rebases, which are quite annoying in such a big PR).

@markus2330 Not sure, how much you actually want to review. Everything should™ be the same as the Python code and documentation you already reviewed. However, I had to make a small change (see also 5839dd4): An escaped key name that is just (namespace +) `/%` is now illegal, because the unescaped version would collide with the root keys (`/`, `user:/` etc.), which already use an empty part.

**Comment by kodebach** [comment]
Not sure what's wrong with the `pdflatex` build (TBH not sure why we even have that build job), but since the macOS jobs are blocked by #3559 I won't fix it for now. Apart from these two build issues the PR is done IMO.

**Review by markus2330 — APPROVED** [review]
Incredible work!

I only looked through https://github.com/ElektraInitiative/libelektra/pull/3555/files/7f767a442377cdb7c80db71cd52f8b3b35e99864..99e0766a86212a3ca0041e7f40165a0341da03c2

Even if someone finds further problems in this PR, I think it is better if we fix them in follow-up PRs. This already worked quite well with @kodebach several times in the past.

So I agree we should merge it before any other PR, @kodebach has already rebased often enough. As master is now broken anyway, we can also merge right away and fix master afterwards. Then it is easier to help @kodebach with the remaining minor issues (like the pdflatex build).

**Comment by kodebach** [comment]
> As master is now broken anyway, we can also merge right away and fix master afterwards. Then it is easier to help @kodebach with the remaining minor issues (like the pdflatex build).

In that case, I will have another look at the `pdflatex` stuff today, but if I can't find a fix quickly we can merge this PR.

**Comment by kodebach** [comment]
@markus I got as far as figuring out that something about `doc/keynames/README.md` is wrong, but not what exactly. If you know about markdown things that break doxygen's latex/pdf generation, please check if you spot any...

**Comment by markus2330** [comment]
Some UTF-8 chars, which are not declared for LaTeX (in doc/markdownlinkconverter/elektraSpecialCharacters.sty) break the pdf build. Maybe it is the (r) sign?

**Comment by kodebach** [comment]
> Some UTF-8 chars, which are not declared for LaTeX (in doc/markdownlinkconverter/elektraSpecialCharacters.sty) break the pdf build. Maybe it is the (r) sign?

Hm, yes that could be that case. It would be an explanation why the build works locally on my machine. Probably differnt LaTeX settings... I'll try replacing it with the HTML entity `&reg;`

**Comment by markus2330** [comment]
You can simply declare ® in doc/markdownlinkconverter/elektraSpecialCharacters.sty, I think this is the better solution.

**Comment by kodebach** [comment]
Seems like it didn't help

**Review by mpranj — CHANGES_REQUESTED** [review]
@kodebach thank you so much for this exceptional work! Thanks as well for fixing all the little problems unrelated to the main topic of this PR. I can't imagine how painful rebasing was, it took me three hours just to scroll throught the changes.

Most importantly:
I think there is an error with clear_bit/test_bit in `src/libs/elektra/keyname.c`.

All other comments are just suggestions, so feel free to ignore the rest.

Additionally, can these empty files be removed:
- `src/bindings/swig/python2/tests/testpy2_kdb.py`
- `doc/todo/PROPOSALS`?

File: doc/VERSION.md
Comment (kodebach) [reply doc/VERSION.md:15]: No that must be a rebase error... :+1:

File: src/libs/elektra/kdb.c
Comment (kodebach) [reply src/libs/elektra/kdb.c:248]: It's been a while, but I'm pretty sure this was obsolete already (Mabye @markus2330 knows more) and I just didn't feel like fixing it.

File: src/plugins/mmapstorage/mmapstorage.c
Comment (kodebach) [reply src/plugins/mmapstorage/mmapstorage.c:760]: Yes, `ukey` and `key` should be allocated and freed together. I even made sure that a `Key` is not modified by e.g. `keySetName` unless the operation will be successful (apart from `malloc` failures).

File: src/bindings/swig/python/tests/test_key.py
Comment (kodebach) [reply src/bindings/swig/python/tests/test_key.py:47]: Because of various changes (foremost the fact that `keyNew (NULL)` now returns `NULL`, i.e. returns an error) `isValid` should now always return true. Also IMO it was pretty weird that `k` was invalid immediately after calling the constructor.

File: doc/help/kdb-meta-show.md
Comment (kodebach) [reply doc/help/kdb-meta-show.md:46]: That's actually unchanged from `master`...

File: src/libs/elektra/keyname.c
Comment (kodebach) [reply src/libs/elektra/keyname.c:473]: Yes, seems to be a copy paste error. Interesting that this didn't cause any memory errors...

File: src/libs/elektra/kdb.c
Comment (kodebach) [reply src/libs/elektra/kdb.c:248]: I just checked and there is a note in the `bootstrap.md` decision that this fallback should be removed with 1.0. That's probably why I removed it instead of updating it:

https://github.com/ElektraInitiative/libelektra/blob/6bb3e2c764870fba0a3299c43264a084a7a1d860/doc/decisions/bootstrap.md#L47-L49

**Comment by kodebach** [comment]
> it took me three hours just to scroll throught the changes.

I'm amazed you did it that quickly...

The `clear_bit` thing was definitely an error. Let's see, if the builds still work. If so, let's merge this PR and fix everything else in a follow up. Since merging this will break anything that depends on `master`, don't think anyone will mind if it also introduces 1 or 2 small bugs ;)

**Review by mpranj — APPROVED** [review]
Excellent work! Again, a huge thank you to @kodebach for making this a reality. 🥇

**Pull request merged**
